File: CS/E4121/Controllers/HomeController.cs
using DevExpress.Data.Filtering;
using DevExpress.Web.Mvc;
using DevExpress.XtraReports.UI;
using E4121.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace E4121.Controllers
{
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            EmailDataGenerator.Register();
            return View();
        }

        [ValidateInput(false)]
        public ActionResult GridViewPartial()
        {
            return PartialView("_GridViewPartial");
        }

        public ActionResult ExportTo(string filterString)
        {
            CriteriaOperator filterCriteria = CriteriaOperator.Parse(filterString);
            IList<Email> emails = LargeDatabaseDataProvider.GetEmailsByFilter(filterCriteria);

            XtraReport1 report = new XtraReport1();
            report.DataSource = emails;
            ExportReport(report, "largeData", true, "Xlsx");
            return GridViewPartial();
        }

        public void ExportReport(XtraReport report, string fileName, bool saveAsFile, string fileFormat)
        {
            using (MemoryStream stream = new MemoryStream())
            {
                report.ExportToXlsx(stream);

                string disposition = saveAsFile ? "attachment" : "inline";

                Response.Clear();
                Response.Buffer = false;
                Response.AppendHeader("Content-Type", string.Format("application/{0}", fileFormat));
                Response.AppendHeader("Content-Transfer-Encoding", "binary");
                Response.AppendHeader("Content-Disposition", string.Format("{0}; filename={1}.{2}", disposition, HttpUtility.UrlEncode(fileName).Replace("+", "%20"), fileFormat));
                Response.BinaryWrite(stream.ToArray());
                Response.End();
            }
        }
    }
}

File: VB/E4121VB/Controllers/HomeController.vb
Imports System.IO
Imports DevExpress.Data.Filtering
Imports DevExpress.Web.Mvc
Imports DevExpress.XtraReports.UI
Public Class HomeController
    Inherits System.Web.Mvc.Controller

    Function Index() As ActionResult
        EmailDataGenerator.Register()
        Return View()
    End Function

    <ValidateInput(False)>
    Public Function GridViewPartial() As ActionResult
        Return PartialView("_GridViewPartial")
    End Function

    Public Function ExportTo(ByVal filterString As String) As ActionResult
        Dim filterCriteria As CriteriaOperator = CriteriaOperator.Parse(filterString)
        Dim emails As IList(Of Email) = LargeDatabaseDataProvider.GetEmailsByFilter(filterCriteria)
        Dim report As XtraReport1 = New XtraReport1()
        report.DataSource = emails
        ExportReport(report, "largeData", True, "Xlsx")
        Return GridViewPartial()
    End Function

    Public Sub ExportReport(ByVal report As XtraReport, ByVal fileName As String, ByVal saveAsFile As Boolean, ByVal fileFormat As String)
        Using stream As MemoryStream = New MemoryStream()
            report.ExportToXlsx(stream)
            Dim disposition As String = If(saveAsFile, "attachment", "inline")
            Response.Clear()
            Response.Buffer = False
            Response.AppendHeader("Content-Type", String.Format("application/{0}", fileFormat))
            Response.AppendHeader("Content-Transfer-Encoding", "binary")
            Response.AppendHeader("Content-Disposition", String.Format("{0}; filename={1}.{2}", disposition, HttpUtility.UrlEncode(fileName).Replace("+", "%20"), fileFormat))
            Response.BinaryWrite(stream.ToArray())
            Response.End()
        End Using
    End Sub
End Class